code/go/0chain.net/smartcontract/storagesc/kill.go

```diff
@@ -0,0 +1,179 @@
+		return conf.OwnerId == t.ClientID
+	}); err != nil {
+		return "", err
+	}
+
+	var req providerRequest
+	if err := req.decode(input); err != nil {
+		return "", common.NewError("kill_blobber_failed", err.Error())
+	}
+
+	var blobber *StorageNode
+	if blobber, err = ssc.getBlobber(req.ID, balances); err != nil {
+		return "", common.NewError("kill_blobber_failed",
+			"can't get the blobber "+t.ClientID+": "+err.Error())
+	}
+	blobber.Kill()
+	if err = emitUpdateBlobber(blobber, balances); err != nil {
+		return "", common.NewError("kill_blobber_failed", err.Error())
+	}
+
+	if blobber.LastRewardPartition.valid() {
+		activePassedBlobberRewardPart, err := getActivePassedBlobberRewardsPartitions(balances, conf.BlockReward.TriggerPeriod)
+		if err != nil {
+			return "", common.NewError("kill_blobber_failed",
+				"cannot get all blobbers list: "+err.Error())
+		}
+		err = activePassedBlobberRewardPart.RemoveItem(balances, blobber.LastRewardPartition.Index, blobber.ID)
+		if err != nil {
+			return "", common.NewError("kill_blobber_failed",
+				"cannot remove blobber from active passed rewards partition: "+err.Error())
+		}
+	}
+
+	if blobber.RewardPartition.valid() {
+		parts, err := getOngoingPassedBlobberRewardsPartitions(balances, conf.BlockReward.TriggerPeriod)
+		if err != nil {
+			return "", common.NewErrorf("commit_connection_failed",
+				"cannot fetch ongoing partition: %v", err)
+		}
+		err = parts.RemoveItem(balances, blobber.RewardPartition.Index, blobber.ID)
+		if err != nil {
+			return "", common.NewError("kill_blobber_failed",
+				"cannot remove blobber from ongoing passed rewards partition: "+err.Error())
+		}
+	}
+
+	var sp *stakePool
+	if sp, err = ssc.getStakePool(blobber.ID, balances); err != nil {
+		return "", common.NewError("update_validator_settings_failed",
+			"can't get related stake pool: "+err.Error())
+	}
+	sp.IsDead = true
+	if err := sp.SlashFraction(
+		conf.StakePool.KillSlash,
+		req.ID,
+		spenum.Blobber,
+		balances,
+	); err != nil {
+		return "", common.NewError("kill_blobber_failed",
+			"can't slash blobber: "+err.Error())
+	}
+	if err = sp.save(ssc.ID, blobber.ID, balances); err != nil {
+		return "", common.NewError("kill_blobber_failed",
+			fmt.Sprintf("saving stake pool: %v", err))
+	}
+
+	if _, err = balances.InsertTrieNode(blobber.GetKey(ssc.ID), blobber); err != nil {
+		return "", common.NewError("kill_blobber_failed",
+			"can't save blobber: "+err.Error())
+	}
+	return "", nil
+}
+
+func (ssc *StorageSmartContract) killValidator(
+	t *transaction.Transaction,
+	input []byte,
+	balances cstate.StateContextI,
+) (string, error) {
+	var err error
+	var conf *Config
+	if conf, err = ssc.getConfig(balances, true); err != nil {
+		return "", common.NewError("kill_validator_failed",
+			"can't get config: "+err.Error())
+	}
+	if err := smartcontractinterface.AuthorizeWithOwner("kill_validator_failed", func() bool {
```

Comment:
probably we can move common code from killBlobber and killValidator to some common func

Reply:
I can have another try, but I am not convinced that this makes the code more readable.